pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,567 @@
+//     that is larger than `b.BarrierTS` (`t' <= max{t, min(b.BarrierTS for b in B)}` where `B` is the set of GC
+//     barriers).
+//
+// TODO: Explicitly state the versions that GCStateManager starts to be functional and the old APIs/concepts/terms is
+//       deprecated when these work are all done.
+
+// GCStateManager is the manager for all kinds of states of TiKV's GC for MVCC data.
+// nolint:revive
+type GCStateManager struct {
+	lock            syncutil.RWMutex
+	gcMetaStorage   endpoint.GCStateProvider
+	cfg             config.PDServerConfig
+	keyspaceManager *keyspace.Manager
+}
+
+// NewGCStateManager creates a GCStateManager of GC and services.
+func NewGCStateManager(store endpoint.GCStateProvider, cfg config.PDServerConfig, keyspaceManager *keyspace.Manager) *GCStateManager {
+	return &GCStateManager{gcMetaStorage: store, cfg: cfg, keyspaceManager: keyspaceManager}
+}
+
+// redirectKeyspace checks the given keyspaceID, and returns the actual keyspaceID to operate on.
+func (m *GCStateManager) redirectKeyspace(keyspaceID uint32, isUserAPI bool) (uint32, error) {
+	// Regard it as NullKeyspaceID if the given one is invalid (exceeds the valid range of keyspace id), no matter
+	// whether it exactly matches the NullKeyspaceID.
+	if keyspaceID & ^constant.ValidKeyspaceIDMask != 0 {
+		return constant.NullKeyspaceID, nil
+	}
+
+	keyspaceMeta, err := m.keyspaceManager.LoadKeyspaceByID(keyspaceID)
+	if err != nil {
+		return 0, err
+	}
+	if keyspaceMeta.Config[keyspace.GCManagementType] != keyspace.KeyspaceLevelGC {
+		if isUserAPI {
+			// The user API is expected to always work. Operate on the state of global GC instead.
+			return constant.NullKeyspaceID, nil
+		}
+		// Internal API should never be called on keyspaces without keyspace level GC. They won't perform any active
+		// GC operation and will be managed by the global GC.
+		return 0, errs.ErrGCOnInvalidKeyspace.GenWithStackByArgs(keyspaceID)
+	}
+
+	return keyspaceID, nil
+}
+
+// AdvanceGCSafePoint tries to advance the GC safe point to the given target. If the target is less than the current
+// value or greater than the txn safe point, it returns an error.
+//
+// WARNING: This method is only used to manage the GC procedure, and should never be called by code that doesn't
+// have the responsibility to manage GC. It can only be called on NullKeyspace or keyspaces with keyspace level GC
+// enabled.
+func (m *GCStateManager) AdvanceGCSafePoint(keyspaceID uint32, target uint64) (oldGCSafePoint uint64, newGCSafePoint uint64, err error) {
+	keyspaceID, err = m.redirectKeyspace(keyspaceID, false)
+	if err != nil {
+		return
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.advanceGCSafePointImpl(keyspaceID, target, false)
+}
+
+func (m *GCStateManager) advanceGCSafePointImpl(keyspaceID uint32, target uint64, compatible bool) (oldGCSafePoint uint64, newGCSafePoint uint64, err error) {
```

Comment:
Where is compatible=true?

Reply:
Not yet included in this PR.
See: https://github.com/tikv/pd/pull/9134